File: hw_01_17_2022184040_sort.py
class City:
  def __init__(self, name, x, y):
    self.name = name
    self.x, self.y = x, y
  def __lt__(self, other):
    return self.name < other.name
  def __repr__(self):
    return '%s(%3d,%3d)' % (self.name, self.x, self.y)

cities = [
  City("Clean", 1336, 536),  City("Prosy", 977, 860),
  City("Rabbi", 6, 758),     City("Addle", 222, 261),
  City("Smell", 1494, 836),  City("Quite", 905, 345),
  City("Lives", 72, 714),    City("Cross", 23, 680),
  City("Synth", 1529, 785),  City("Tweak", 1046, 426),
  City("Medic", 1485, 514),  City("Glade", 660, 476),
  City("Breve", 1586, 448),  City("Hotel", 1269, 576),
  City("Toing", 398, 561),   City("Scorn", 617, 373),
  City("Tweet", 1253, 403),  City("Zilch", 1289, 29),
  City("React", 296, 659),   City("Fiche", 787, 278),
]

print(cities)

# name asc
cities.sort()       # mutable
# ss = sorted(cities) # immutable
print(cities)

# x desc
cities.sort(key=lambda c:-c.x)
print(cities)

# y asc
cities.sort(key=lambda c:c.y)
print(cities)

File: hw_01_17_2022184040_sort.java
// online java compiler:
//   https://codechacha.com/ko/java-sorting-array/
//   https://www.onlinegdb.com/

import java.util.Arrays;
import java.util.Comparator;
import java.lang.Comparable;

public class Main {
  public static class City implements Comparable<City> {
    String name; int x, y;
    public City(String name, int x, int y) {
      this.name = name;
      this.x = x;
      this.y = y;
    }
    public int compareTo(City other) {
      return this.name.compareTo(other.name);
    }
    public String toString() {
      return name+'('+x+','+y+')';
    }
  }
  static City[] cities = new City[] {
    new City("Clean", 1336, 536),    new City("Prosy", 977, 860),
    new City("Rabbi", 6, 758),       new City("Addle", 222, 261),
    new City("Smell", 1494, 836),    new City("Quite", 905, 345),
    new City("Lives", 72, 714),      new City("Cross", 23, 680),
    new City("Synth", 1529, 785),    new City("Tweak", 1046, 426),
    new City("Medic", 1485, 514),    new City("Glade", 660, 476),
    new City("Breve", 1586, 448),    new City("Hotel", 1269, 576),
    new City("Toing", 398, 561),     new City("Scorn", 617, 373),
    new City("Tweet", 1253, 403),    new City("Zilch", 1289, 29),
    new City("React", 296, 659),     new City("Fiche", 787, 278),
  };
  public static void main(String args[]) {
    System.out.println(Arrays.toString(cities));
    // sort here by name
    // Arrays.sort(cities, new Comparator<City>() {
    //     public int compare(City a, City b) {
    //         return a.name.compareTo(b.name);
    //     }
    // });
    Arrays.sort(cities);
    System.out.println(Arrays.toString(cities));
    // x coordinate desc
    Arrays.sort(cities, new Comparator<City>() {
        public int compare(City a, City b) {
            return b.x - a.x;
        }
    });
    System.out.println(Arrays.toString(cities));
    // y coordinate asc
    Arrays.sort(cities, new Comparator<City>() {
        public int compare(City a, City b) {
            return a.y - b.y;
        }
    });
    System.out.println(Arrays.toString(cities));
  }
}
